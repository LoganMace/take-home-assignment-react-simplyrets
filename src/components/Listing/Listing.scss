.listing {
  width: 315px;

  &__image-wrap {
    height: 280px;
    width: 315px;
    overflow: hidden;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 4px;
    margin-bottom: 16px;
    position: relative;
  }

  &__image {
    height: 280px;
  }

  &__fav-icon {
    position: absolute;
    top: 18px;
    right: 24px;
    height: 36px;
    width: 39px;
    cursor: pointer;
  }

  &__interior {
    font-size: 20px;
    font-weight: 600;
  }

  &__price {
    font-size: 26px;
    font-weight: 700;
  }

  &__address {
    font-size: 15px;
    font-weight: 400;
  }

  &__date {
    color: #979797;
    font-size: 14px;
    font-weight: 400;
  }
}