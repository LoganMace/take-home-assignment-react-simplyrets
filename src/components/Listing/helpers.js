export const formatPrice = (price) => {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    maximumFractionDigits: 0,
    minimumFractionDigits: 0,
  }).format(price);
};

export const formatDate = (date) => {
  const baseDate = new Date(date);
  const month = baseDate.getMonth() + 1;
  const day = baseDate.getDate();
  const year = baseDate.getFullYear().toString().substring(2, 4);
  const dateString = `${month}/${day}/${year}`;

  return dateString;
};

export const calculateBaths = (fullBaths, halfBaths) => {
  const totalBaths = fullBaths + (halfBaths * .5);

  return totalBaths;
};
