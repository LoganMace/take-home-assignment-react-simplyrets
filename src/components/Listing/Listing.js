import React, { useEffect, useState } from 'react';
import heartFill from '../../assets/heart-fill.svg';
import heartStroke from '../../assets/heart-stroke.svg';
import './Listing.scss';
import { formatDate, formatPrice, calculateBaths } from './helpers';

export const Listing = ({
  property,
  isFavorite,
  toggleFavorite,
  id
}) => {
  const {
    photo,
    bedrooms,
    bathsFull,
    bathsHalf,
    area,
    listPrice,
    address: {
      streetNumber,
      streetName,
      city,
      state
    },
    listDate
  } = property;

  const [isSaved, setIsSaved] = useState(isFavorite)

  useEffect(() => {
    setIsSaved(isFavorite)
  }, [isFavorite]);

  const getHeartIcon = () => (
    <div data-testid={`favorite-icon-${id}`} onClick={() => toggleFavorite(id)}>
      <img
        src={isSaved ? heartFill : heartStroke}
        className="listing__fav-icon"
        alt="favorite icon"
      />
    </div>
  );

  return (
    <div data-testid={`property-listing`} className='listing'>
      <div className='listing__image-wrap'>
        <img className='listing__image' src={photo} alt='property listing image' />
        {getHeartIcon()}
      </div>
      <div className='listing__interior'>{bedrooms} BR | {calculateBaths(bathsFull, bathsHalf)} Bath | {area} Sq Ft</div>
      <div className='listing__price'>{formatPrice(listPrice)}</div>
      <div className='listing__address'>{streetNumber} {streetName}, {city}, {state}</div>
      <div className='listing__date'>Listed: {formatDate(listDate)}</div>
    </div>
  );
}
