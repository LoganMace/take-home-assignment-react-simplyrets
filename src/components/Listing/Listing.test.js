import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { Listing } from './Listing';

const mockProperty = {
  photo: 'test-image',
  bedrooms: 4,
  bathsFull: 3,
  bathsHalf: 2,
  area: 3456,
  listPrice: 1234567,
  address: {
    streetNumber: 404,
    streetName: 'Main St',
    city: 'Dallas',
    state: 'TX'
  },
  listDate: '12/03/2020'
};

describe('Listing Component', () => {
  test('should render', () => {
    const { asFragment } = render(
      <Listing property={mockProperty} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should show filled heart', () => {
    const { asFragment } = render(
      <Listing property={mockProperty} isFavorite={true} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should call toggleFavorite on click', () => {
    const toggleMock = jest.fn();
    const { asFragment, getByTestId } = render(
      <Listing property={mockProperty} toggleFavorite={toggleMock} id='1' />
    );
    const favIcon = getByTestId('favorite-icon-1');
    fireEvent.click(favIcon);
    expect(toggleMock).toHaveBeenCalledWith('1');
    expect(asFragment()).toMatchSnapshot();
  });
});
