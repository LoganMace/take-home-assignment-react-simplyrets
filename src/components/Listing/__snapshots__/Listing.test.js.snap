// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listing Component should call toggleFavorite on click 1`] = `
<DocumentFragment>
  <div
    class="listing"
    data-testid="property-listing"
  >
    <div
      class="listing__image-wrap"
    >
      <img
        alt="house front"
        class="listing__image"
        src="test-image"
      />
      <div
        data-testid="favorite-icon-1"
      >
        <img
          alt="favorite icon"
          class="listing__fav-icon"
          src="heart-stroke.svg"
        />
      </div>
    </div>
    <div
      class="listing__interior"
    >
      4 BR | 4 Bath | 3456 Sq Ft
    </div>
    <div
      class="listing__price"
    >
      $1,234,567
    </div>
    <div
      class="listing__address"
    >
      404 Main St, Dallas, TX
    </div>
    <div
      class="listing__date"
    >
      Listed: 12/3/20
    </div>
  </div>
</DocumentFragment>
`;

exports[`Listing Component should render 1`] = `
<DocumentFragment>
  <div
    class="listing"
    data-testid="property-listing"
  >
    <div
      class="listing__image-wrap"
    >
      <img
        alt="house front"
        class="listing__image"
        src="test-image"
      />
      <div
        data-testid="favorite-icon-undefined"
      >
        <img
          alt="favorite icon"
          class="listing__fav-icon"
          src="heart-stroke.svg"
        />
      </div>
    </div>
    <div
      class="listing__interior"
    >
      4 BR | 4 Bath | 3456 Sq Ft
    </div>
    <div
      class="listing__price"
    >
      $1,234,567
    </div>
    <div
      class="listing__address"
    >
      404 Main St, Dallas, TX
    </div>
    <div
      class="listing__date"
    >
      Listed: 12/3/20
    </div>
  </div>
</DocumentFragment>
`;

exports[`Listing Component should show filled heart 1`] = `
<DocumentFragment>
  <div
    class="listing"
    data-testid="property-listing"
  >
    <div
      class="listing__image-wrap"
    >
      <img
        alt="house front"
        class="listing__image"
        src="test-image"
      />
      <div
        data-testid="favorite-icon-undefined"
      >
        <img
          alt="favorite icon"
          class="listing__fav-icon"
          src="heart-fill.svg"
        />
      </div>
    </div>
    <div
      class="listing__interior"
    >
      4 BR | 4 Bath | 3456 Sq Ft
    </div>
    <div
      class="listing__price"
    >
      $1,234,567
    </div>
    <div
      class="listing__address"
    >
      404 Main St, Dallas, TX
    </div>
    <div
      class="listing__date"
    >
      Listed: 12/3/20
    </div>
  </div>
</DocumentFragment>
`;
